packages/editor/src/components/post-actions/set-as-homepage.js

```diff
@@ -0,0 +1,217 @@
+			currentHomePage: homePage.postId,
+		};
+	} );
+	const [ currentHomePageTitle, setCurrentHomePageTitle ] = useState( '' );
+	const isPageDraft = item.status === 'draft';
+
+	const { getEntityRecord } = resolveSelect( coreStore );
+	const { saveEditedEntityRecord, saveEntityRecord } =
+		useDispatch( coreStore );
+	const { createSuccessNotice, createErrorNotice } =
+		useDispatch( noticesStore );
+
+	useEffect( () => {
+		const handleGetCurrentHomepage = async () => {
+			const currentHomePageItem = await getEntityRecord(
+				'postType',
+				'page',
+				currentHomePage
+			);
+			setCurrentHomePageTitle( getItemTitle( currentHomePageItem ) );
+		};
+		handleGetCurrentHomepage();
+	}, [ currentHomePage, getEntityRecord ] );
+
+	async function onSetPageAsHomepage( event ) {
+		event.preventDefault();
+
+		try {
+			// If selected page is set to draft, publish the page.
+			if ( isPageDraft ) {
+				await saveEntityRecord( 'postType', 'page', {
+					...item,
+					status: 'publish',
+				} );
+			}
+
+			// Save new home page settings.
+			await saveEntityRecord( 'root', 'site', {
+				page_on_front: item.id,
+				show_on_front: 'page',
+			} );
+
+			closeModal?.();
+
+			await saveEditedEntityRecord( 'root', 'site', undefined, {
+				page_on_front: item.id,
+				show_on_front: 'page',
+			} );
+
+			createSuccessNotice( __( 'Homepage updated' ), {
+				type: 'snackbar',
+			} );
+
+			onActionPerformed?.( items );
+		} catch ( error ) {
+			const typedError = error;
+			const errorMessage =
+				typedError.message && typedError.code !== 'unknown_error'
+					? typedError.message
+					: __( 'An error occurred while setting the homepage' );
+			createErrorNotice( errorMessage, { type: 'snackbar' } );
+		}
+	}
+
+	const renderModalBody = () => {
+		if ( 'posts' === showOnFront ) {
+			return (
+				<>
+					<Text>
+						{ sprintf(
+							// translators: %s: title of the page to be set as the homepage.
+							__(
+								'Set "%s" as the site homepage? This will replace the current homepage which is set to display latest posts.'
+							),
+							pageTitle
+						) }
+					</Text>
+				</>
+			);
+		}
+
+		if ( isPageDraft ) {
+			return (
+				<>
+					<Text>
+						{ sprintf(
+							// translators: %s: title of the page to be set as the homepage.
+							__(
+								'"%s" is a draft and will be published automatically if set as the homepage. Publish page and set as the site homepage?'
+							),
+							pageTitle
+						) }
+					</Text>
+				</>
+			);
+		}
+
+		const modalTranslatedString =
+			// translators: %1$s: title of page to be set as the home page. %2$s: title of the current home page.
+			__(
+				'Set "%1$s" as the site homepage? This will replace the current homepage: "%2$s"'
+			);
+
+		return (
+			<Text>
+				{ sprintf(
+					modalTranslatedString,
+					pageTitle,
+					currentHomePageTitle
+				) }{ ' ' }
+			</Text>
+		);
+	};
+
+	return (
+		<form onSubmit={ onSetPageAsHomepage }>
```

Comment:
Yes, nice idea! I've added the `isSavingEntityRecord` selector and disabled the modal button when saving is in progress: https://github.com/WordPress/gutenberg/pull/65426/commits/b0297944108936df021a72d80508d4272268caea.

Reply:
I think this is not longer relevant now that we don't allow for draft Pages to be set as the homepage?

Reply:
> I think this is not longer relevant now that we don't allow for draft Pages to be set as the homepage?

Yes that's right, for now, but it will likely need to be handled in the follow-up that handles draft pages.